### PR TITLE
Reuse the cache created for latest `main` on PRs/branches if `setup.py` is not modified

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

For a PR or a branch, if `setup.py` is not modified (compare to the common ancestor with the `main`), let's use cache that is created for the `setup.py` from the `latest` commit on the `main` branch. 

**latest** means the latest one on `main` at the moment where a run is triggered.

Motivation:

  - avoid having cache for `main` and `pull` (for most cases), which is introduced in #24886 to avoid unexpected/undesired edge cases.
  - **avoid the cache creation and store for PRs/branches that still have (very) old `setup.py` as they don't rebase on more recent `main`**

With this PR, we expect the storage usage for `cache` could be reduced dramatically 🤞 .

[The artifact in this job run](https://app.circleci.com/pipelines/github/huggingface/transformers/70312/workflows/b93db7c8-40ac-4925-ac2a-ea9feac9b926/jobs/881950/artifacts) shows `generated_config.txt` has explicit checksum used `v0.7-pipelines_torch-main-pip-9RXs1YQ8L2beP4cdAfRDkWX0VRTtWaQodDVKzvyJwPI=` rather than `{{ checksum "setup.py" }}`

### CODE REVIEW

File: src/transformers/models/bert/modeling_bert.py [hunk 1]
```diff
@@ -883,7 +883,7 @@ class BertModel(BertPreTrainedModel):
     to `True`. To be used in a Seq2Seq model, the model needs to initialized with both `is_decoder` argument and
     `add_cross_attention` set to `True`; an `encoder_hidden_states` is then expected as an input to the forward pass.
     """
-
+    a = 3
```
Comment (ydshieh) [inline src/transformers/models/bert/modeling_bert.py:886]: I will remove this before merge

File: .circleci/create_circleci_config.py [hunk 2]
```diff
@@ -536,6 +542,19 @@ def create_circleci_config(folder=None):
...
+    if os.path.exists(test_file):
+        with open(test_file) as f:
+            test_list = f.read()
+    else:
+        test_list = []
+
+    use_cache_on_main = False
+    # `setup.py` is not modified
+    if test_list != "tests":
+        use_cache_on_main = True
```
Comment (ydshieh) [inline .circleci/create_circleci_config.py:556]: Use the cache from `main` if `setup.py` is not modified.

Two advantages:
  
  - we don't need anymore to keep 2 cache (as currently): one for `main` and one for PRs/branches
  - we

File: src/transformers/models/bert/modeling_bert.py [hunk 3]
```diff
@@ -884,6 +884,8 @@ class BertModel(BertPreTrainedModel):
     `add_cross_attention` set to `True`; an `encoder_hidden_states` is then expected as an input to the forward pass.
     """
 
+    a = 3
```
Comment (ydshieh) [inline src/transformers/models/bert/modeling_bert.py:887]: I will remove this before merge

File: .circleci/create_circleci_config.py [hunk 4]
```diff
@@ -100,14 +106,22 @@ def to_dict(self):
             job["resource_class"] = self.resource_class
         if self.parallelism is not None:
             job["parallelism"] = self.parallelism
+
+        checksum = self.checksum if self.checksum is not None else '{{ checksum "setup.py" }}'
+        save_cache = True
+        if self.checksum is not None:
+            # `setup.py` is not modified and we are not on `main` branch
```
Comment (ydshieh) [inline .circleci/create_circleci_config.py:113]: `self.checksum` is not None only if:
  - the triggered event is not from `main` branch
  - `setup.py` is not modified

File: .circleci/create_circleci_config.py [hunk 5]
```diff
@@ -100,14 +106,22 @@ def to_dict(self):
...
+            cache_branch_prefix = "main"
+            save_cache = False
```
Comment (ydshieh) [inline .circleci/create_circleci_config.py:115]: always avoid writing to cache from a PR that could potentially used by `main`

File: .circleci/create_circleci_config.py [hunk 6]
```diff
@@ -531,11 +546,49 @@ def job_name(self):
...
+
+    main_head.checkout()
+    proc = subprocess.Popen(["sha256sum", f"{setup_file_path}"], stdout=subprocess.PIPE)
+    checksum = proc.stdout.read().decode().split(" ")[0]
+    checksum = b64encode(bytes.fromhex(checksum)).decode()
+
+    # go back to the original branch
+    current_head.checkout()
+
+    return checksum
```
Comment (ydshieh) [inline .circleci/create_circleci_config.py:567]: It's probably better to record this information, so we can debug (check which one is used) easily when something goes wrong

**Review by sgugger — APPROVED** [review]
Thanks for working on this! Let's try it for a bit and see if it impacts any workflow negatively.

File: .circleci/config.yml [hunk 7]
```diff
@@ -34,9 +34,9 @@ jobs:
             - run: pip install -U --upgrade-strategy eager GitPython
             - run: pip install -U --upgrade-strategy eager .
             - run: mkdir -p test_preparation
-            - run: python utils/tests_fetcher.py | tee tests_fetched_summary.txt
+            - run: python utils/tests_fetcher.py | tee test_preparation/tests_fetched_summary.txt
             - store_artifacts:
-                  path: ~/transformers/tests_fetched_summary.txt
+                  path: test_preparation/tests_fetched_summary.txt
```
Comment (ydshieh) [inline .circleci/config.yml:39]: need this in `create_circleci_config.py` to check if `tests` is in the list

**Comment by ydshieh** [comment]
@sgugger FYI: I need to change 2 more places to make it work correctly for all cases. See my last 2 review comments.

**Review by sgugger — APPROVED** [review]
Works for me!

**Comment by ydshieh** [comment]
Hi @sgugger 

I am afraid I have to revert this PR. [From this section](
https://circleci.com/docs/caching/#caching-and-open-source)

> PRs from the same fork repo share a cache (this includes, as previously stated, that PRs in the main repo share a cache with main).
> Two PRs in different fork repos have different caches.

The cache is never shared between PRs/branches from different forks. This might explains the question I posted about why the same cache key could be found sometimes but not other times. 

The PR description of #24886 is partially valid as **lhoestq** created a branch in `transformers` rather than a PR from his own forked repo.

The way to share cache is

> Enabling the sharing of[ environment variables](https://circleci.com/docs/env-vars/) allows cache sharing between the original repo and all forked builds.

But we need to be careful if we have sensitive env. variables or not.

**Pull request merged**